### PR TITLE
Wake AIX up from coma

### PR DESCRIPTION
**Pull request by NattyNarwhal**
This gets OTP building on AIX and able to run readymade Elixir builds; support was present, but subtly broken. I plan to check if some other stuff works (RabbitMQ and Pleroma make good real-world applications on my radar.) but this is a surprisingly far start for little work.

Points ranging from what I do on my environment to non-specific AIX-specific tweaks:

* I fixed the AIX checks in the configure script to work for modern AIX, and work on PASE. (I haven't tested this, but it's next on my agenda.)

* 32-bit build seems broken; it causes the assembler to throw up. Not sure why; you can set a 64-bit build with `OBJECT_MODE=64 CC="gcc -maix64"` during configure (because GCC doesn't respect the magic environment variable AIX uses for bitness of toolchain, grr) and `OBJECT_MODE=64` during make.

* You need `_ALL_SOURCE` for build to get the structure the ioctl for changing window size uses to appear. Should place in configure script, not sure where.

* I had GNU binutils on my `PATH`; this is a bad idea because it sucks on AIX. 

* Likewise, OpenSSL linking stuff I had to point explicitly to `/opt/freeware` (the prefix for third-party RPMs on AIX) because some decrepit IBM supplied version is in `/usr` it seems.

* Some of the patches were adapted from Bull's RPM specfile for OTP R15.

* Documentation build is untested, as is Java and WxWidgets support.

* I'm amazed all the platform-specific code for AIX still works after all these years...

### CODE REVIEW

**Comment by NattyNarwhal** [comment]
Some other points I forgot:

* The BEAM definition of `PRIORITY_MAX` overrides the AIX definition, which is different. This doesn't seem to be harming anything, and would require some copious sed abuse to fix.

* Native libraries (i.e: `erl_interface`) only seem to be building in the static objects. This is a problem for packages with native dependencies. This is likely because of AIX hairiness with shared objects. (tl;dr: shared objects usually live inside of ar archives)

File: erts/emulator/sys/unix/erl_child_setup.c [hunk 1]
```diff
@@ -491,7 +491,12 @@ main(int argc, char *argv[])
             ErtsSysForkerProto proto;
             errno = 0;
             if ((res = sys_uds_read(uds_fd, (char*)&proto, sizeof(proto),
+#if defined(MSG_DONTWAIT)
                                     pipes, 3, MSG_DONTWAIT)) < 0) {
+#else
+		/* XXX: a workaround here? */
```
Comment (kadler) [inline erts/emulator/sys/unix/erl_child_setup.c]: AIX has `MSG_NONBLOCK`, which is equivalent. Just use something like:

```C
#if !defined(MSG_DONTWAIT) && defined(MSG_NONBLOCK)
#define MSG_DONTWAIT MSG_NONBLOCK
#endif
```

**Comment by aboroska** [comment]
Regarding `erl_interface` I suggest to define `LINUX_SOURCE_COMPAT` [1]

Note that the C standard allows the `malloc(0)` system call to return either
zero or a pointer that cannot be dereferenced. Unixes typically return a
non-zero pointer, but AIX returns NULL unless one defines `LINUX_SOURCE_COMPAT`.

User of legacy drivers might experience crash [2] here when they convert an empty list/string to binary:
https://github.com/erlang/otp/blob/OTP-21.2.4/lib/erl_interface/src/legacy/erl_eterm.c#L800
because `erl_malloc` expects the pointer here:
https://github.com/erlang/otp/blob//OTP-21.2.4/lib/erl_interface/src/legacy/erl_malloc.c#L234

[1] CFLAGS="-D_LINUX_SOURCE_COMPAT"
[2] "ERROR erl_malloc: Failed to allocate more memory"

**Comment by jhogberg** [comment]
Thanks for the PR!

> ...  and I had to nop out a function set in ex_doc, otherwise I get a bad error that I'm not sure where it'd be coming from:

This was caused by a compiler bug that affected all platforms, it was fixed earlier today in 2b603d86dbc8715eda4efc78294ded0770d2149e.

**Comment by garazdawi** [comment]
We're aiming to get this PR into OTP-22-rc3. We were just about to test it, but realized that is conflicts with changes in our internet test branch. Once they are merged we will ask you to rebase ontop of latest master and then we'll include this in testing.

**Comment by NattyNarwhal** [comment]
Resolved merge conflicts and fixed up indentation a little as a result. Were these the conflicts you were talking about?

**Comment by garazdawi** [comment]
Yes, perfect. Thank you.

File: lib/erl_interface/include/ei.h [hunk 2]
```diff
@@ -329,7 +329,7 @@ typedef struct {
     int (*socket)(void **ctx, void *setup_ctx);
     int	(*close)(void *ctx);
     int (*listen)(void *ctx, void *addr, int *len, int backlog);
-    int (*accept)(void **ctx, void *addr, int *len, unsigned tmo);
```
Comment (garazdawi) [inline lib/erl_interface/include/ei.h]: Just noticed this change.... this is a backward incompatible change, which means that there needs to be a very good reason to change it, and I'm not sure AIX support qualifies.

Not sure how to work around it except only renaming the field on AIX, though that is not a very pretty solution.

Comment (NattyNarwhal) [reply lib/erl_interface/include/ei.h]: There is the possibility of `#undefine accept` but then that gets crumbly once you try using system accept calls.

I could try renaming it back and do an `#ifdef` approach though.

**Comment by NattyNarwhal** [comment]
Sorry for the force push, I had pulled latest master and rebased onto it, but git thought I was still behind.

The stupidity regarding this whole renaming accept-et-al charade comes from AIX's `socket.h`. Peer into this fractal of WTF yourself:

```c
#ifdef _ALL_SOURCE
/*
 * In a non-SPEC 1170 environment, we'd rather just call the "n" versions of
 * these routines, because they're a little bit quicker.
 */
#ifndef _KERNEL
#ifdef __cplusplus
extern "C" int naccept(int, struct sockaddr *, socklen_t *);
extern "C" int ngetpeername(int, struct sockaddr *, socklen_t *);
extern "C" int ngetsockname(int, struct sockaddr *, socklen_t *);
extern "C" ssize_t nrecvfrom(int, void *, size_t, int, struct sockaddr *,
                             socklen_t *);
inline int accept(int socketfd, struct sockaddr *address, socklen_t *address_len)
{
    return naccept(socketfd, address, address_len);
}
inline int getpeername(int socketfd, struct sockaddr *address,
                       socklen_t *address_len)
{
    return ngetpeername(socketfd, address, address_len);
}
inline int getsockname(int socketfd, struct sockaddr *address,
                       socklen_t *address_len)
{
    return ngetsockname(socketfd, address, address_len);
}
inline ssize_t recvfrom(int socketfd, void *buffer, size_t length, int flags,
                        struct sockaddr *address, socklen_t *address_len)
{
    return nrecvfrom(socketfd, buffer, length, flags, address,
                 address_len);
}
int getpeereid(int, uid_t *__restrict__, gid_t *__restrict__);
#ifndef COMPAT_43
extern "C" ssize_t nrecvmsg(int, struct msghdr *, int);
extern "C" ssize_t nsendmsg(int, const struct msghdr *, int);
inline ssize_t recvmsg(int socketfd, struct msghdr *message, int flags)
{
    return nrecvmsg(socketfd, message, flags);
}
inline ssize_t sendmsg(int socketfd, const struct msghdr *message, int flags)
{
    return nsendmsg(socketfd, message, flags);
}
#else /* COMPAT_43 */

extern "C" ssize_t recvmsg(int, struct msghdr *, int);
extern "C" ssize_t sendmsg(int, const struct msghdr *, int);

#endif /* COMPAT_43 */
#else  /* __cplusplus */

#define recvfrom        nrecvfrom
#define accept          naccept
#define getpeername     ngetpeername
#define getsockname     ngetsockname

#ifndef COMPAT_43
#define recvmsg         nrecvmsg
#define sendmsg         nsendmsg

#endif /* COMPAT_43 */
#endif /* __cplusplus */

#endif /* _KERNEL */
// more...
```

File: lib/erl_interface/include/ei.h (hunk 2 above)
Comment (garazdawi) [reply lib/erl_interface/include/ei.h]: thanks for the super quick response. I'll put this in test over the weekend and then merge.

**Pull request merged**